rustls/src/compress.rs

```diff
@@ -75,3 +81,67 @@ pub struct DecompressionFailed;
 /// A content-less error for when `CertCompressor::compress` fails.
 #[derive(Debug)]
 pub struct CompressionFailed;
+
+#[cfg(feature = "zlib-rs")]
+mod feat_zlib_rs {
+    use zlib_rs::{deflate, inflate, ReturnCode};
+
+    use super::*;
+
+    /// A certificate decompressor for the Zlib algorithm using the `zlib-rs` crate.
+    #[derive(Debug)]
+    pub struct ZlibRsDecompressor;
+
+    impl CertDecompressor for ZlibRsDecompressor {
+        fn algorithm(&self) -> CertificateCompressionAlgorithm {
+            CertificateCompressionAlgorithm::Zlib
+        }
+
+        fn decompress(&self, input: &[u8], output: &mut [u8]) -> Result<(), DecompressionFailed> {
+            let (output_filled, rc) =
+                inflate::uncompress_slice(output, input, inflate::InflateConfig::default());
+            if rc != ReturnCode::Ok {
+                return Err(DecompressionFailed);
+            }
+
+            if output_filled.len() != output.len() {
+                return Err(DecompressionFailed);
+            }
+
+            Ok(())
+        }
+    }
+
+    /// A certificate compressor for the Zlib algorithm using the `zlib-rs` crate.
+    #[derive(Debug)]
+    pub struct ZlibRsCompressor;
+
+    impl CertCompressor for ZlibRsCompressor {
+        fn algorithm(&self) -> CertificateCompressionAlgorithm {
+            CertificateCompressionAlgorithm::Zlib
+        }
+
+        fn compress(
+            &self,
+            input: Vec<u8>,
+            level: CompressionLevel,
+        ) -> Result<Vec<u8>, CompressionFailed> {
+            let mut output = alloc::vec![0u8; deflate::compress_bound(input.len())];
+            let config = match level {
+                CompressionLevel::Interactive => deflate::DeflateConfig::default(),
+                CompressionLevel::Amortized => deflate::DeflateConfig::new(9),
+            };
+            let (output_filled, rc) = deflate::compress_slice(&mut output, &input, config);
+            if rc != ReturnCode::Ok {
+                return Err(CompressionFailed);
+            }
+
+            let used = output_filled.len();
+            output.truncate(used);
```

Comment:
nit: I think the `used` binding doesn't add much here. Inline?

Reply:
Realized after the fact this is load bearing, it prevents two mutable borrows of `output` :+1: